### PR TITLE
adding summary stat workflow

### PR DESCRIPTION
**Pull request by andrewkern**
Adding a separate workflow to compute summary statistics along simulated chromosomes. 

The basic flow is tree sequence -> vcf -> `diploSHIC`, which I use to compute windowed summary statistics from each of the simulated chrms / seeds / populations / etc

The outputs are in a new directory produced by the workflow called `results/summaries` and the major products are `csv` files with the windowed stats summarized across seeds for each model/dfe/annotation/population/chrm combination along with plots of a subset of those stats -- currently pi,  tajD, and ZnS

The directory structure of the output is as so for a given configuration with two seeds
```

results/summaries/OutOfAfrica_3G09
├── Gamma_H17
│   └── ensembl_havana_104_exons
│       ├── 1675701734
│       ├── 1845187043
│       ├── sim_chr21.allseeds.CEU.diploshic.stats
│       ├── sim_chr21.allseeds.CEU.pi.png
│       ├── sim_chr21.allseeds.CHB.diploshic.stats
│       ├── sim_chr21.allseeds.CHB.pi.png
│       ├── sim_chr21.allseeds.YRI.diploshic.stats
│       └── sim_chr21.allseeds.YRI.pi.png
└── none
    └── none
        ├── 1675701734
        ├── 1845187043
        ├── sim_chr21.allseeds.CEU.diploshic.stats
        ├── sim_chr21.allseeds.CEU.pi.png
        ├── sim_chr21.allseeds.CHB.diploshic.stats
        ├── sim_chr21.allseeds.CHB.pi.png
        ├── sim_chr21.allseeds.YRI.diploshic.stats
        └── sim_chr21.allseeds.YRI.pi.png
```

and here is an example plot from that showing two replicates from the two seeds for the CHB sample

sim_chr21 allseeds CHB pi

currently this PR could probably use some polishing, but I won't have time for it in the next few days, so perhaps others can lay eyes on this?

### CODE REVIEW

**Comment by andrewkern** [comment]
tagging you all @nspope @mufernando @stsmall  -- hoping to get some eyes on this PR for calculating summary stats using diploshic

**Comment by stsmall** [comment]
OK. I am running through this now.

**Comment by andrewkern** [comment]
yeah agreed. that would be a good addition.

**Comment by andrewkern** [comment]
i also realized that this is computing "diploid" stats now and we should probably show "haploid" stats instead. i'm going to push a quick commit while you're looking it over

File: workflows/summary_stats.snake [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+rule make_vcf:
+    input:
+        output_dir + "/simulated_data/{demog}/{dfes}/{annots}/{seeds}/sim_{chrms}.trees"
+
+    output:
+        output_dir + "/summaries/{demog}/{dfes}/{annots}/{seeds}/sim_{chrms}.vcf"
+    run:
+        ts = tskit.load(input[0])
+        ts.write_vcf(open(output[0], "w"))
```
Comment (stsmall) [inline workflows/summary_stats.snake:121]: I think this would be cleaner, since then I am sure the file is closed, as 
```
with open(output[0], 'w') as vcf:
    ts.write_vcf(vcf, contig_id=wildcards.chrms)
```
If you add the contig_id then you can cat the diploshic output files for each population in a seed. Then when plotting set the chromosome as a hue or in relplot as rows.

**Comment by stsmall** [comment]
I played around with plotting. Esp useful would be a boxplot of the exons and intron for the different DFEs. I think it would be an easier check on whether diversity was being reduced. These ended up being some big changes, mainly because I cat all the diploshic files together across seeds and DFEs. It may be better as a branch after the PR is accepted.

**Comment by stsmall** [comment]
Everything else looked OK. Hard to tell from a pi landscape what was really happening and why I think a box plot would be a good addition.

**Comment by chriscrsmith** [comment]
Related to what stsmall was saying, here's neutral and selected overlaid (chr 21 CEU I think):
[pi_plot.pdf](https://github.com/popsim-consortium/analysis2/files/10841355/pi_plot.pdf)

**Comment by stsmall** [comment]
> 

if you have the code handy ... can you add the exon locations w/ rug

**Comment by chriscrsmith** [comment]
boxplot, chr 21 CEU 
[bp.pdf](https://github.com/popsim-consortium/analysis2/files/10841449/bp.pdf)

**Comment by stsmall** [comment]
@chriscrsmith NICE!!!
I was expecting a more dramatic diff

**Comment by andrewkern** [comment]
@chriscrsmith is that fig one replicate each? I've got summary stats for chr13 for 10 reps at this point IIRC

**Comment by chriscrsmith** [comment]
I ran with tiny_config which says reps=2. I don't think I'm seeing the 2 reps though?

Looking closely at the 100 chunks—most .stats files are empty. Only a few chunks have output. Does that sound right?

**Comment by andrewkern** [comment]
that's right. some of those files will be empty. at the end of the day you should get a file structure like:
```
results/summaries/OutOfAfrica_3G09/Gamma_H17/ensembl_havana_104_exons
├── 1358822686
├── 137773603
├── 1402032511
├── 1675701734
├── 170765738
├── 1845187043
├── 290876774
├── 561383554
├── 789925285
├── 878579711
├── sim_chr13.allseeds.CEU.diploshic.stats
├── sim_chr13.allseeds.CEU.pi.png
├── sim_chr13.allseeds.CHB.diploshic.stats
├── sim_chr13.allseeds.CHB.pi.png
├── sim_chr13.allseeds.YRI.diploshic.stats
└── sim_chr13.allseeds.YRI.pi.png
```

where the `*.allseeds.*.stats` files have all the collected stats across seeds and the `*.png` files have summary stat looks like http://poppy.uoregon.edu/~adkern/stdpopsim/results/summaries/OutOfAfrica_3G09/Gamma_H17/ensembl_havana_104_exons/sim_chr13.allseeds.CEU.pi.png

**Comment by andrewkern** [comment]
I just had to push a little addition to this as pandas didn't like having lots of seeds

**Comment by chriscrsmith** [comment]
averaged two reps. looking much better! (running more reps now)
[pi_plot.pdf](https://github.com/popsim-consortium/analysis2/files/10842163/pi_plot.pdf)
[bp.pdf](https://github.com/popsim-consortium/analysis2/files/10842164/bp.pdf)

File: workflows/summary_stats.snake (hunk 1 above)
Comment (stsmall) [reply workflows/summary_stats.snake:121]: if you add this change you also have to add wildcards.chrms to the ancestralAllelesFile so it assoc w/ the correct ancestral file. However, then there is another error. 
```
Traceback (most recent call last):                                                                                                                                                                                 
  File "/home/ssmall/mambaforge/envs/analysis2/bin/makeFeatureVecsForChrArmFromVcf_ogSHIC.py", line 129, in <module>                                                                                               
    genos = allel.GenotypeArray(genos.subset(sel0=snpIndicesToKeep))                                                                                                                                               
  File "/home/ssmall/mambaforge/envs/analysis2/lib/python3.10/site-packages/allel/model/ndarray.py", line 1530, in subset                                                                                          
    return subset_genotype_array(self, sel0, sel1, cls=type(self), subset=subset)                                                                                                                                  
  File "/home/ssmall/mambaforge/envs/analysis2/lib/python3.10/site-packages/allel/model/generic.py", line 227, in subset_genotype_array                                                                            
    out = subset(g.values, sel0, sel1, **kwargs)                                                                                                                                                                   
  File "/home/ssmall/mambaforge/envs/analysis2/lib/python3.10/site-packages/allel/model/ndarray.py", line 69, in subset                                                                                            
    return data[sel0, sel1]                                                                                                                                                                                        
IndexError: arrays used as indices must be of integer (or boolean) type 
```

unless this is an easy fix, maybe not worth recording the correct chrom name in the diploshic output?

Comment (andrewkern) [reply workflows/summary_stats.snake:121]: yeah let's hold off and get this PR merged first.

**Comment by andrewkern** [comment]
It would be great to get this merged this week. where is everyone with this PR?

**Comment by chriscrsmith** [comment]
> 

I say merge!

**Comment by nspope** [comment]
Sorry @andrewkern, I'm slammed this week -- haven't had time for a look.

**Comment by mufernando** [comment]
I gave it a high level look and it looked OK to me. Since this PR is self contained I see no reason not to merge it whenever?

**Comment by stsmall** [comment]
As is, it works as expected. It is self contained and I didnt get any errors. It is not currently a rule included in all, were you going to add it? It does take a fairly long time to run.
tldr; merge away

**Comment by andrewkern** [comment]
i was not going to add a rule to the main workflow, but we could later. merging this now then and we can keep refining.
